### DISCUSSION TITLE
Is it possible to override the name of the model used in the JSON Schema?

### DISCUSSION THREAD

**Discussion by bobbyrullo** [pydantic/pydantic#2467] [Question]
Hi there,

We have a pretty large project that we are moving from a custom `attrs` based solution to Pydantic. As such, we have a number of conflicting names. eg:

```python
# foo.py

class Reference(BaseModel):
    """A reference to a Foo"""
    ...

# bar.py
class Reference(BaseModel):
    """A reference to a Bar"""
    ...
```

So in Python-land we have a `bar.Reference` and a `foo.Reference`. But JSON Schema for an OpenAPI set of definitions is one flat namespace. Pydantic handles this automatically but I do not like the results, since it looks like "mod__other_mod__Foo". 

Is there a way to customize the name used in definitions? I tried overriding Config.title but that doesn't override the key used in the OpenAPI `#definitions`. 

I could of course rename my classes, but I'd rather not since then we'd have the repetitive `foo.FooReference` and `bar.BarRefernence`

Thanks!

Bobby

**Comment by PrettyWood** [comment]
Hi @bobbyrullo 
When two models have the same name,
https://github.com/samuelcolvin/pydantic/blob/619ff261c9bdc2a612ef2fcfa17a005dfefbcf21/pydantic/schema.py#L416-L417
is called.
If you just want to change the name of the model, you could just change `__name__` temporary like this
```py
from contextlib import contextmanager

from pydantic import BaseModel


@contextmanager
def change_cls_name(cls, tmp_cls_name):
    original_cls_name = cls.__name__
    cls.__name__ = tmp_cls_name
    yield
    cls.__name__ = original_cls_name


class Reference(BaseModel):
    """A reference to a Bar"""

    @classmethod
    def schema(cls, *a, **kw):
        with change_cls_name(cls, 'Pika'):
            return super().schema(*a, **kw)

assert Reference.schema()['title'] == 'Pika'
assert Reference.__name__ == 'Reference'
```

**Reply by bobbyrullo** [reply]
Ha, that's not bad. I'd rather have something a bit more supported, as this relies too much on knowledge of pydantic internals. Still, could be a decent workaround in the meantime.

Thanks @PrettyWood

**Reply by PrettyWood** [reply]
A tiny PR could maybe be done like this
```diff
diff --git a/pydantic/schema.py b/pydantic/schema.py
index 4dc47f0..7abe9e7 100644
--- a/pydantic/schema.py
+++ b/pydantic/schema.py
@@ -317,7 +317,7 @@ def get_model_name_map(unique_models: TypeModelSet) -> Dict[TypeModelOrEnum, str
     name_model_map = {}
     conflicting_names: Set[str] = set()
     for model in unique_models:
-        model_name = normalize_name(model.__name__)
+        model_name = getattr(model, '__pydantic_schema_name__', None) or normalize_name(model.__name__)
         if model_name in conflicting_names:
             model_name = get_long_model_name(model)
             name_model_map[model_name] = model
```
```py
class Model(BaseModel):
    __pydantic_schema_name__ = 'Pika'

    x: str
```
@samuelcolvin WDTY?

**Comment by PrettyWood** [comment]
You can always also use `Config.title` as described in the [doc](https://pydantic-docs.helpmanual.io/usage/model_config/).
And now (since v1.8) that you can specify config options as model class kwargs it's even shorter
```py
class Reference(BaseModel, title='bar.Reference'):
    """A reference to a Bar"""

assert Reference.schema() == {
    'title': 'bar.Reference',
    'description': 'A reference to a Bar',
    'type': 'object',
    'properties': {},
}
```

**Reply by bobbyrullo** [reply]
I tried that but the key in the schema dict is the problem, it's still 'Reference':

```python

In [4]: schema.schema({Reference})
Out[4]:
{'definitions': {'Reference': {'title': 'bar.Reference',
   'description': 'A reference to a Bar',
   'type': 'object',
   'properties': {}}}}

```

**Comment by s3i7h** [comment]
I was about to make an issue about this and found this page.

~I will use the above fix for now (thank you @PrettyWood!), but I think the naming of a model should be manipulatable, and I was thinking something like the proposed fix.~ 
The fix didn't work :( 
`get_model_name_map()` doesn't trigger `cls.schema()` thus I couldn't hook in to it. I had to permanently override `cls.__name__` with something like:

```
def classname(__cls_name__: str):
    def inner(cls: type):
        cls.__name__ = __cls_name__
        return cls

    return inner

@classname("B")
class A(BaseModel):
    ...
```

I was trying to change the `Config.title` hoping it would change the name too, and found out it doesn't. I couldn't find out the details about the `title` field in a JSON schema, but isn't it almost equal to the name?
I was thinking in that case, `Config.title` might be a candidate for the model name also. Are there any insights here?

**Comment by nmoreaud** [comment]
@bobbyrullo How did you solve your problem?

Here is another workaround:
```python
# foo.py
class ReferenceFoo(BaseModel):
    """A reference to a Foo"""
    ...

Reference = ReferenceFoo

# bar.py
class ReferenceBar(BaseModel):
    """A reference to a Bar"""
    ...

Reference = ReferenceBar
```
